src/hamiltonian.jl

```diff
@@ -41,12 +41,16 @@ function ∂H∂θ(h::Hamiltonian, θ::AbstractVecOrMat)
     return DualValue(res[1], -res[2])
 end
 
-∂H∂r(h::Hamiltonian{<:UnitEuclideanMetric,<:GaussianKinetic}, r::AbstractVecOrMat) = copy(r)
-function ∂H∂r(h::Hamiltonian{<:DiagEuclideanMetric,<:GaussianKinetic}, r::AbstractVecOrMat)
-    return h.metric.M⁻¹ .* r
+∂H∂r(h::Hamiltonian{<:UnitEuclideanMetric,<:GaussianKinetic}, r::AbstractArray) = copy(r)
+function ∂H∂r(h::Hamiltonian{<:DiagEuclideanMetric,<:GaussianKinetic}, r::AbstractArray)
+    out = similar(r)
+    out .= h.metric.M⁻¹ .* r
+    return out
```

Comment:
According to the issue only `DenseEuclideanMetric` was problematic. Was this also broken? If not, the current code is actually better IMO - it avoids changing from e.g. `SMatrix` to `MMatrix` and the new code would be broken for vectors with element types that the result can't be written into (probably e.g. `Vector{Int}`). So generally IMO the old code is better if it works.

Reply:
Yeah, in the diagonal case, since the mass matrix is only represented using a Vector, so the original way is better